fdbclient/VersionVector.h

```diff
@@ -29,30 +29,52 @@
 #include "fdbclient/FDBTypes.h"
 #include "fdbclient/Knobs.h"
 
+static const int InvalidEncodedSize = 0;
+
 struct VersionVector {
-	boost::container::flat_map<Tag, Version> versions;
+	boost::container::flat_map<Tag, Version> versions; // An ordered map. (Note:
+	                                                   // changing this to an unordered
+	                                                   // map will break the
+	                                                   // serialization code below.)
 	Version maxVersion; // Specifies the max version in this version vector. (Note:
 	                    // there may or may not be a corresponding entry for this
 	                    // version in the "versions" map.)
 
-	VersionVector() : maxVersion(invalidVersion) {}
-	VersionVector(Version version) : maxVersion(version) {}
+	VersionVector() : maxVersion(invalidVersion), cachedEncodedSize(InvalidEncodedSize) {}
+	VersionVector(Version version) : maxVersion(version), cachedEncodedSize(InvalidEncodedSize) {}
 
 private:
 	// Only invoked by getDelta() and applyDelta(), where tag has been validated
 	// and version is guaranteed to be larger than the existing value.
-	inline void setVersionNoCheck(const Tag& tag, Version version) { versions[tag] = version; }
+	inline void setVersionNoCheck(const Tag& tag, Version version) {
+		versions[tag] = version;
+		invalidateCachedEncodedSize();
+	}
+
+	inline void invalidateCachedEncodedSize() { cachedEncodedSize = InvalidEncodedSize; }
+
+	// Encoded version vector size. Introduced to help speed up serialization.
+	// @note This encoded size is not meant to be kept in sync with the updates
+	// that happen to the version vector.
+	// @note A value of 0 (= InvalidEncodedSize) indicates that the encoded version
+	// vector size is not cached.
+	size_t cachedEncodedSize;
 
 public:
 	Version getMaxVersion() const { return maxVersion; }
 
+	void setMaxVersion(Version version) { maxVersion = version; }
+
 	int size() const { return versions.size(); }
 
+	bool empty() const { return versions.empty(); }
+
 	void setVersion(const Tag& tag, Version version) {
 		ASSERT(tag != invalidTag);
```

Comment:
can you check the tag locality is not equal to tagLocalityInvalid ? And maybe check if it should be less than tagLocalityInvalid.

Is there an upper bound to tag.id  we can check? i.e. # storage servers?

Reply:
> can you check the tag locality is not equal to tagLocalityInvalid ? And maybe check if it should be less than tagLocalityInvalid.
> 

Done.

> Is there an upper bound to tag.id we can check? i.e. # storage servers?

I don't know if we can make any assumptions about max tag id value based on number of storage servers, and the storage server count information may not be available anyway.  Why do we want to check this?